xdsl/transforms/accfg/trace_acc_state.py

```diff
@@ -0,0 +1,74 @@
+"""
+These inference helpers are used to figure out which values are
+present at which points in the IR.
+
+Example inference results:
+
+```
+accfg.setup(A = %1, B = %2)  // previous state = {}
+// ...
+accfg.setup(A = %3, C = %1)  // previous state = {A = %1, B = %2}
+```
+
+These inference passes walk the IR backwards.
+"""
+
+from xdsl.dialects import accfg, scf
+from xdsl.ir import Block, SSAValue
+
+State = dict[str, SSAValue]
+
+
+def infer_state_of(state_var: SSAValue) -> State:
+    """
+    Entrance function of the inference pass.
+
+    This walks up the def-use chain to compute all values
+    that are guaranteed to be set in this state.
+    """
+    owner = state_var.owner
+    match owner:
+        # If the owner is a setup op that has no previous state,
+        # return all values of the previous state as an op.
+        case accfg.SetupOp(in_state=None) as setup_op:
+            return {name: val for name, val in setup_op.iter_params()}
+        # If the owner is a setup op with a previous state,
+        # update the state with all the values declared there.
+        case accfg.SetupOp(in_state=st) as setup_op if st is not None:
+            in_state = infer_state_of(st)
+            in_state.update(dict(setup_op.iter_params()))
+            return in_state
+        # If the owner is an scf.if that has two possible setups
+        # only update the state that is set in both possible setups
+        case scf.If() as if_op:
+            return state_union(*infer_states_for_if(if_op, state_var))
+        case Block():
+            # TODO: maybe implement something here?
+            return {}
+        case _:
```

Comment:
@AntonLydike, do I get it correctly that this is only an issue if at some point we lower to cf instead of working on scf? Returning an empty dict feels wrong here (can we also just throw an error here ? I don't understand why pyright is not complaining actually, is this not a dict{None: None} that  is returned?

Reply:
Yes, or if we want to handle state transfers across function calls at some point. But that's far, far away.